### PR TITLE
Nats protobuf

### PR DESCRIPTION
**Pull request by afanasievM**
added protobuf to nats
@ajax-gnatiuk-s

### CODE REVIEW

File: university-service/src/testFixtures/kotlin/validation/ValidationTestFixture.kt [hunk 1]
```diff
@@ -0,0 +1,12 @@
+package validation
+
+import javax.validation.Validation
+import javax.validation.Validator
+
+open class ValidationTestFixture {
```
Comment (ajax-gnatiuk-s) [inline university-service/src/testFixtures/kotlin/validation/ValidationTestFixture.kt]: why the class is open?

File: transaction-service/src/main/kotlin/ua/com/foxminded/restClient/nats/NatsController.kt [hunk 2]
```diff
@@ -35,20 +38,36 @@ class NatsController @Autowired constructor(
 
 
     fun handleMessage(message: Message) {
-        println(message.data.decodeToString())
-        val request = mapper.readValue<TransactionNatsDto>(message.data.decodeToString())
-        val transactions = findTransactions(
-            request.id!!, request.startDate!!, request.endDate,
-            Pageable.ofSize(request.size!!).withPage(request.page!!)
+        val request = ProtoMessage.TransactionRequestProto.newBuilder().mergeFrom(message.data).build()
```
Comment (ajax-gnatiuk-s) [inline transaction-service/src/main/kotlin/ua/com/foxminded/restClient/nats/NatsController.kt]: Would it work
```kotlin
ProtoMessage.TransactionRequestProto.parseFrom(message.data)
```

?

File: transaction-service/src/main/kotlin/ua/com/foxminded/restClient/nats/NatsController.kt [hunk 3]
```diff
@@ -35,20 +38,36 @@ class NatsController @Autowired constructor(
...
+    private fun mapTransactionDtoToResponse(dto: TransactionDto): ProtoMessage.TransactionResponseProto {
+        return ProtoMessage.TransactionResponseProto.newBuilder()
+            .setId(dto.id.toString())
+            .setPersonId(dto.personId.toString())
+            .setTransactionTime(
+                Timestamp.newBuilder()
+                    .setSeconds(dto.transactionTime!!.toEpochSecond(ZoneOffset.UTC))
+                    .setNanos(dto.transactionTime!!.nano)
+            )
+            .setValue(dto.value!!)
```
Comment (ajax-gnatiuk-s) [inline transaction-service/src/main/kotlin/ua/com/foxminded/restClient/nats/NatsController.kt]: Could we avoid `!!`?

**Review by ajax-gnatiuk-s — COMMENTED** [review]
NATS controller looks good to me. It would be nice to have an integration test that sends requests to the controller

File: university-service/src/testFixtures/kotlin/validation/ValidationTestFixture.kt (hunk 1 above)
Comment (afanasievM) [reply university-service/src/testFixtures/kotlin/validation/ValidationTestFixture.kt]: it's redundant
fixed

File: transaction-service/src/main/kotlin/ua/com/foxminded/restClient/nats/NatsController.kt (hunk 2 above)
Comment (afanasievM) [reply transaction-service/src/main/kotlin/ua/com/foxminded/restClient/nats/NatsController.kt]: Yes, it's work. Thank you

File: transaction-service/src/main/kotlin/ua/com/foxminded/restClient/nats/NatsController.kt (hunk 3 above)
Comment (afanasievM) [reply transaction-service/src/main/kotlin/ua/com/foxminded/restClient/nats/NatsController.kt]: I will try. Look at next commit)

File: university-service/src/testFixtures/kotlin/validation/ValidationTestFixture.kt (hunk 1 above)
Comment (ajax-gnatiuk-s) [reply university-service/src/testFixtures/kotlin/validation/ValidationTestFixture.kt]: @afanasievM 

So, we found out that `open` is redundant. The next step is making the class `object` or omit class declaration at all.

We don't need companion object :)

```kotlin 
object ValidationTestFixture { // bad class name. It's better to use something like `Validations`
    val validator: Validator = Validation.buildDefaultValidatorFactory().validator 
}
```

File: transaction-service/src/test/kotlin/ua/com/foxminded/restClient/NatsControllerTest.kt [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+            .setPersonId("e966f608-4621-11ed-b878-0242ac120002")
+            .setCurrency("UAH")
+            .setStartDate(Timestamp.newBuilder().setSeconds(startDate.epochSecond).setNanos(startDate.nano).build())
+            .setEndDate(Timestamp.newBuilder().setSeconds(endDate.epochSecond).setNanos(endDate.nano).build())
+            .setPage(0)
+            .setSize(2)
+            .build()
+    }
+
+    fun getListTransactions(): List<TransactionDto> {
```
Comment (ajax-gnatiuk-s) [inline transaction-service/src/test/kotlin/ua/com/foxminded/restClient/NatsControllerTest.kt]: `private`?

File: transaction-service/src/test/kotlin/ua/com/foxminded/restClient/NatsControllerTest.kt [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+    fun setUp() {
+        nats = GenericContainer("nats:latest").withExposedPorts(NATS_PORT, NATS_MGMT_PORT)
+        nats.start()
+        natsUrl = "nats://%s:%d".format(nats.host, nats.getMappedPort(NATS_PORT))
+        natsConnection = Nats.connect(natsUrl)
+        natsController = NatsController(transactionService, exchangeService, natsConnection)
+    }
+
+    @Test
+    fun nats_shouldReturnFluxTransactions_whenInputCorrect() {
```
Comment (ajax-gnatiuk-s) [inline transaction-service/src/test/kotlin/ua/com/foxminded/restClient/NatsControllerTest.kt]: you can rename with 
```kotlin
fun `nats should return flux ...`() {
```

It's much more easier to read

File: transaction-service/src/test/kotlin/ua/com/foxminded/restClient/NatsControllerTest.kt (hunk 4 above)
Comment (afanasievM) [reply transaction-service/src/test/kotlin/ua/com/foxminded/restClient/NatsControllerTest.kt]: +

File: transaction-service/src/test/kotlin/ua/com/foxminded/restClient/NatsControllerTest.kt (hunk 5 above)
Comment (afanasievM) [reply transaction-service/src/test/kotlin/ua/com/foxminded/restClient/NatsControllerTest.kt]: okay.

File: university-service/src/testFixtures/kotlin/validation/ValidationTestFixture.kt [hunk 6]
```diff
@@ -0,0 +1,12 @@
+package validation
+
+import javax.validation.Validation
+import javax.validation.Validator
+
+class ValidationTestFixture {
+
+    companion object {
```
Comment (ajax-gnatiuk-s) [inline university-service/src/testFixtures/kotlin/validation/ValidationTestFixture.kt]: please look at this [comment](https://github.com/afanasievM/rest-project/pull/9#discussion_r1318225598).

We can simplify the class by using `object`

File: transaction-service/src/test/kotlin/ua/com/foxminded/restClient/NatsControllerTest.kt [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+        val startDate = LocalDateTime.parse("2022-11-01T12:00:00").atZone(ZoneId.systemDefault()).toInstant()
+        val endDate = LocalDateTime.parse("2022-11-01T12:00:00").atZone(ZoneId.systemDefault()).toInstant()
+        return ProtoMessage.TransactionRequestProto.newBuilder()
+            .setPersonId("e966f608-4621-11ed-b878-0242ac120002")
+            .setCurrency("UAH")
+            .setStartDate(Timestamp.newBuilder().setSeconds(startDate.epochSecond).setNanos(startDate.nano).build())
+            .setEndDate(Timestamp.newBuilder().setSeconds(endDate.epochSecond).setNanos(endDate.nano).build())
+            .setPage(0)
+            .setSize(2)
+            .build()
```
Comment (ajax-gnatiuk-s) [inline transaction-service/src/test/kotlin/ua/com/foxminded/restClient/NatsControllerTest.kt]: There are some tricks how to make the code simpler:

```suggestion
        return ProtoMessage.TransactionRequestProto.newBuilder().apply {
            personId = "e966f608-4621-11ed-b878-0242ac120002"
            currency = "UAH"
            startDateBuilder.setSeconds(startDate.epochSecond).setNanos(startDate.nano)   
            endDateBuilder.setSeconds(endDate.epochSecond).setNanos(endDate.nano)
            page = 0
            size = 2
         }.build()
```

It's not call to action, just sharing knowledge :)

**Review by ajax-gnatiuk-s — APPROVED** [review]

File: university-service/src/testFixtures/kotlin/validation/ValidationTestFixture.kt (hunk 6 above)
Comment (afanasievM) [reply university-service/src/testFixtures/kotlin/validation/ValidationTestFixture.kt]: ok. 
I don't know why I miss it(

File: transaction-service/src/test/kotlin/ua/com/foxminded/restClient/NatsControllerTest.kt (hunk 7 above)
Comment (afanasievM) [reply transaction-service/src/test/kotlin/ua/com/foxminded/restClient/NatsControllerTest.kt]: Ok. I just try commit suggestion on GitHub)

**Pull request merged**